# Lab 9: Imputing Missing Values
## IBM Data Analytics Capstone Project 
### Objective: Identify, impute missing values and normalize data in the Stack Overflow Developer Survey dataset

In [1]:
import pandas as pd

# Load the dataset
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/n01PQ9pSmiRX6520flujwQ/survey-data.csv"
df = pd.read_csv(file_path)

print("Dataset loaded successfully!")
print("Shape:", df.shape)
df.head()

Dataset loaded successfully!
Shape: (65437, 114)


,ResponseId,MainBranch,Age,Employment,RemoteWork,Check,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,...,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,SurveyLength,SurveyEase,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,Under 18 years old,"Employed, full-time",Remote,Apples,Hobby,Primary/elementary school,Books / Physical media,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I am a developer by profession,35-44 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
2,3,I am a developer by profession,45-54 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Easy,NaN,NaN
3,4,I am learning to code,18-24 years old,"Student, full-time",NaN,Apples,NaN,Some college/university study without earning ...,"Other online resources (e.g., videos, blogs, f...",Stack Overflow;How-to videos;Interactive tutorial,...,NaN,NaN,NaN,NaN,NaN,NaN,Too long,Easy,NaN,NaN
4,5,I am a developer by profession,18-24 years old,"Student, full-time",NaN,Apples,NaN,"Secondary school (e.g. American high school, G...","Other online resources (e.g., videos, blogs, f...",Technical documentation;Blogs;Written Tutorial...,...,NaN,NaN,NaN,NaN,NaN,NaN,Too short,Easy,NaN,NaN


In [3]:
# Task 1: Identify duplicate rows
print("Number of duplicate rows:", df.duplicated().sum())

# Task 2: Remove duplicates
df = df.drop_duplicates()
print("Shape after removing duplicates:", df.shape)
print("Duplicates removed!")

Number of duplicate rows: 0
Shape after removing duplicates: (65437, 114)
Duplicates removed!


In [4]:
# Task 3: Find missing values for all columns
missing_values = df.isnull().sum()
missing_percent = (df.isnull().sum() / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage (%)': missing_percent.round(2)
}).sort_values('Missing Values', ascending=False)

print("Top 10 columns with most missing values:")
print(missing_df[missing_df['Missing Values'] > 0].head(10))

# Task 4: Missing rows in RemoteWork column
print("\nMissing values in RemoteWork:", df['RemoteWork'].isnull().sum())

Top 10 columns with most missing values:
                               Missing Values  Percentage (%)
AINextMuch less integrated              64289           98.25
AINextLess integrated                   63082           96.40
AINextNo change                         52939           80.90
AINextMuch more integrated              51999           79.46
EmbeddedAdmired                         48704           74.43
EmbeddedWantToWorkWith                  47837           73.10
EmbeddedHaveWorkedWith                  43223           66.05
ConvertedCompYearly                     42002           64.19
AIToolNot interested in Using           41023           62.69
AINextMore integrated                   41009           62.67

Missing values in RemoteWork: 10631


In [5]:
# Task 5: Value counts for RemoteWork
print("RemoteWork value counts:")
print(df['RemoteWork'].value_counts())

# Task 6: Most frequent value
most_frequent_remotework = df['RemoteWork'].value_counts().idxmax()
print("\nMost frequent RemoteWork value:", most_frequent_remotework)

RemoteWork value counts:
RemoteWork
Hybrid (some remote, some in-person)    23015
Remote                                  20831
In-person                               10960
Name: count, dtype: int64

Most frequent RemoteWork value: Hybrid (some remote, some in-person)


In [6]:
# Task 7: Impute missing values in RemoteWork with majority value
df['RemoteWork'] = df['RemoteWork'].fillna(most_frequent_remotework)

print("Missing values after imputation:", df['RemoteWork'].isnull().sum())
print("RemoteWork imputed successfully!")
print("\nUpdated value counts:")
print(df['RemoteWork'].value_counts())

Missing values after imputation: 0
RemoteWork imputed successfully!

Updated value counts:
RemoteWork
Hybrid (some remote, some in-person)    33646
Remote                                  20831
In-person                               10960
Name: count, dtype: int64


In [7]:
# Task 8: Check compensation-related columns
comp_cols = ['CompTotal', 'ConvertedCompYearly']

for col in comp_cols:
    print(f"\n{col}:")
    print(f"  Missing values: {df[col].isnull().sum()}")
    print(f"  Description:\n{df[col].describe()}")


CompTotal:
  Missing values: 31697
  Description:
count     3.374000e+04
mean     2.963841e+145
std      5.444117e+147
min       0.000000e+00
25%       6.000000e+04
50%       1.100000e+05
75%       2.500000e+05
max      1.000000e+150
Name: CompTotal, dtype: float64

ConvertedCompYearly:
  Missing values: 42002
  Description:
count    2.343500e+04
mean     8.615529e+04
std      1.867570e+05
min      1.000000e+00
25%      3.271200e+04
50%      6.500000e+04
75%      1.079715e+05
max      1.625660e+07
Name: ConvertedCompYearly, dtype: float64


## Summary

### Tasks Completed:
- **Duplicates:** 0 found, dataset already clean
- **Missing Values:** 109 out of 114 columns had missing data
- **RemoteWork:** 10,631 missing values imputed with majority value 
  → "Hybrid (some remote, some in-person)"
  → Count increased from 23,015 to 33,646

### Compensation Analysis:
- **CompTotal:** 31,697 missing values, contains extreme outliers (max = 1e+150)
  → Not reliable for analysis
- **ConvertedCompYearly:** 42,002 missing values, cleaner distribution
  → Median: $65,000 | Mean: $86,155 | Max: $16,255,600
  → Preferred column for compensation analysis

### Key Decision:
Use **ConvertedCompYearly** over CompTotal for any salary analysis
as it is normalized and free of extreme outliers.